ui/index.js

```diff
@@ -30,16 +30,38 @@ import txHelper from './helpers/utils/tx-helper';
 
 log.setLevel(global.METAMASK_DEBUG ? 'debug' : 'warn');
 
+let reduxStore;
+
+/**
+ * Method to update backgroundConnection object use by UI
+ *
+ * @param backgroundConnection - connection object to background
+ */
+export const updateBackgroundConnection = (backgroundConnection) => {
+  actions._setBackgroundConnection(backgroundConnection);
+  backgroundConnection.onNotification((data) => {
+    if (data.method === 'sendUpdate') {
+      reduxStore.dispatch(actions.updateMetamaskState(data.params[0]));
+    } else {
+      throw new Error(
+        `Internal JSON-RPC Notification Not Handled:\n\n ${JSON.stringify(
+          data,
+        )}`,
+      );
+    }
+  });
+};
+
 export default function launchMetamaskUi(opts, cb) {
   const { backgroundConnection } = opts;
-  actions._setBackgroundConnection(backgroundConnection);
```

Comment:
We no longer have to call this here?

Reply:
This is now done [here](https://github.com/MetaMask/metamask-extension/pull/14781/files#diff-bb36c8e9593a9516a59e8710c563c14ae46a1a58515a5b7859233da7a0868ba3R165).

Reply:
Could you explain a bit what moving this accomplished?

Reply:
It is just code cleanup, as this functionality is now invoked from 2 places.

Reply:
As I tested it seems to work well.

Reply:
Were you able to test hitting [this if block](https://github.com/MetaMask/metamask-extension/blob/051254dffdcaacbb8cb7639f5bcd537a85cc255a/ui/index.js#L107)?

Reply:
yep I did and it seems to work well

Reply:
I agree it ooks brittle.

- backgroundConnection.getState calls its callback, which calls startApp, which is asynchronous and reduxStore is only set in a .then
- meanwhile, synchronously after, so before reduxStore is set, updateBackgroundConnection(backgroundConnection); is called

It will only work if we can guarantee startApp always resolves before backgroundConnection.onNotification event is triggered.  

Also, startApp makes 2 calls to actions, which means startApp can use actions but it's secretly limited to only the actions that don't depend on background connection.

I'd recommend keeping the additional early call and inside the _setBackgroundConnection function do nothing if argument is `===` to what's already set.

Reply:
As discussed in code walkthrough a race condition is unlikely here